Support function_call in `autogen/agent`

kevin666aa opened the pull request:
<!-- Thank you for your contribution! Please review https://microsoft.github.io/FLAML/docs/Contribute before opening a pull request. -->

<!-- Please add a reviewer to the assignee section when you create a PR. If you don't have the access to it, we will shortly find a reviewer and assign them to your PR. -->

## Why are these changes needed?

Preliminary designs to add the new function_call feature into the agent framework.
<!-- Please give a short summary of the change and the problem this solves. -->

## Related issue number

See #1083 for tasks.
<!-- For example: "Closes #1234" -->

## Checks

<!-- - I've used [pre-commit](https://microsoft.github.io/FLAML/docs/Contribute#pre-commit) to lint the changes in this PR (note the same in integrated in our CI checks). -->
- [x] I've included any doc changes needed for https://microsoft.github.io/FLAML/. See https://microsoft.github.io/FLAML/docs/Contribute#documentation to build and test documentation locally.
- [x] I've added tests (if relevant) corresponding to the changes introduced in this PR.
- [ ] I've made sure all auto checks have passed.

qingyun-wu commented on flaml/autogen/agent/agent.py:
```diff
@@ -30,22 +31,73 @@ def _remember(self, memory):
         It needs to be overriden by the subclass to perform followup actions.
+
+        Args:
+            message (dict or str): message from the sender. If the type is dict, it can contain at most 4 fields:
```
Revise docstr. E.g., "it can contain at most 4 fields" is not correct.

sonichi reviewed (commented):
merge with main and simplify the notebook.
add test.

kevin666aa commented on flaml/autogen/agent/user_proxy_agent.py:
```diff
@@ -90,23 +97,110 @@ def _execute_code(self, code_blocks):
+        result = []
+        inside_quotes = False
+        for char in jstr:
+            if char == '"':
```
Yes, `\\"` can be allowed inside quotes with this implementation, check test_func_call line 76. 
Notebook is revised and test is added. But the notebook and test are work in process and need more refinements.

kevin666aa replied:
Yes this is not right. I have revised it, thanks!

kevin666aa commented:
Some Notes:
1. In `UserProxyAgent.receive`, I added back the line to convert the message to a dict. This is to make the type of the message consistent to avoid keep checking the type. Thus,  We may define `is_termination_msg` to always take in a dict. The type of message passed to `auto_reply` should also be a dict.

2. `UserProxyAgent` now allows the function calls calling to a class method, and passing in predefined arguments, or even update these arguments if needed. The current design is to ask the user to pass in a dict. More details in documentation of `UserProxyAgent`.

sonichi reviewed (commented):
I haven't reviewed the notebook autogen_agent_function_call. Please let me know if it's ready for review.

kevin666aa commented on flaml/autogen/agent/user_proxy_agent.py:
```diff
@@ -136,5 +265,5 @@ def receive(self, message, sender):
 
         self._consecutive_auto_reply_counter[sender.name] += 1
-        print("\n>>>>>>>> NO HUMAN INPUT RECEIVED. USING AUTO REPLY FOR THE USER...", flush=True)
+        print(">>>>>>>> NO HUMAN INPUT RECEIVED. USING AUTO REPLY FOR THE USER...", flush=True)
```
got it!

kevin666aa commented on flaml/autogen/agent/user_proxy_agent.py:
```diff
@@ -33,19 +36,40 @@ def __init__(
                     the number of auto reply reaches the max_consecutive_auto_reply.
                 (3) When "NEVER", the agent will never prompt for human input. Under this mode, the conversation stops
                     when the number of auto reply reaches the max_consecutive_auto_reply or when is_termination_msg is True.
+            functions (dict[str, dict]): a dictionary of dictionaries.
```
Got it!

kevin666aa commented on flaml/autogen/agent/user_proxy_agent.py:
```diff
@@ -90,23 +121,121 @@ def _execute_code(self, code_blocks):
         Once a message is received, this function sends a reply to the sender or simply stop.
         The reply can be generated automatically or entered manually by a human.
         """
+        if type(message) is str:
+            message = {"content": message, "role": "user"}
```
What do you mean by "created function"?

sonichi replied:
I mean assuming you will create a function in Agent, use that function.

kevin666aa replied:
I add a function `_message_to_dict` for this operation.

kevin666aa commented on flaml/autogen/agent/user_proxy_agent.py:
```diff
@@ -33,19 +36,40 @@ def __init__(
+                                index of the return: name of argument to be updated with this returned value in "args".
+                    }
+                Caution: arguments stored in 'args' will overwritten any given arguments passed back from the LLM assistant.
+                You may choose to not pass the argument field to the LLM assistant, or tell the assistant to ignore the argument.
```
I added the docs. I revised the design and remove the 'args' and 'args_to_update' field. It is much easier for one to just define a class to store any arguments.

sonichi commented on flaml/autogen/agent/math_user_proxy_agent.py:
```diff
@@ -84,6 +84,10 @@
 def is_termination_msg(x):
     """Check if a message is a termination message."""
+    if isinstance(x, dict):
+        if x.get("content", None) is None:
+            return False
+        x = x["content"]
```
```suggestion
        x = x.get("content")
        if x is None:
            return False
```

sonichi commented on flaml/autogen/agent/math_user_proxy_agent.py:
```diff
@@ -150,11 +155,20 @@ def __init__(
+                        "class" (Required): an instance of a class.
+                        "func_name" (Optional, str): name of the function in the class. If not given the class will be called directly.
+                    }
+                    See the examples in docstr of UserProxyAgent for more details.
```
"docstr of UserProxyAgent" -> a link

sonichi commented on flaml/autogen/agent/user_proxy_agent.py:
```diff
@@ -33,19 +36,61 @@ def __init__(
+                    }
+
+                Example 1:
+                def add_num_func(num_to_be_added):
```
Use code blocks. Follow existing docstr format. Compile the doc and verify it looks good.

sonichi commented on flaml/autogen/agent/user_proxy_agent.py:
```diff
@@ -33,19 +36,61 @@ def __init__(
+                    name="test",
+                    function_map={
+                        "add_num": {
+                            "class": AddNum(given_num=10),
```
From this example, the value for "class" is actually an instance, not a class.
One thing I don't understand is why not also use `"function": AddNum(given_num=10).add`?

kevin666aa replied:
You are right, there are no need for such a complex feature at all. I will simply this part.

kevin666aa replied:
I have changed this back to previous implementation. User can just pass in `UserProxyAgent(function_map = {"func_name": callable_func})`, similarly `UserProxyAgent(function_map = {"func_name": class.callable_func})`.

sonichi commented on flaml/autogen/agent/user_proxy_agent.py:
```diff
@@ -33,19 +36,61 @@ def __init__(
-            is_termination_msg if is_termination_msg is not None else (lambda x: x == "TERMINATE")
+            is_termination_msg
+            if is_termination_msg is not None
+            else (lambda x: x == "TERMINATE" if isinstance(x, str) else x.get("content") == "TERMINATE")
```
x is a dict.

sonichi commented on website/docs/Use-Cases/Auto-Generation.md:
````diff
@@ -424,6 +424,47 @@ In the example above, we create an AssistantAgent named "assistant" to serve as
+user._execute_function(func_call) # this will call AddNumClass.add_num_func(5) and return 20. The same class instance is used, and the state is preserved.
+```
+
+The code above is used to demonstrate the function calling feature of `UserProxyAgent`. In practice, the function call is triggered by the assistant agent. The assistant agent will send a message to the user proxy agent, which contains the function call information. The user proxy agent will then call the function and return the result to the assistant agent. Please find practical examples in the ...
````
It'll be hard to find that example far away.

kevin666aa replied:
I am rewind changes to Auto-Generation.md and change it in a separate PR if needed.

sonichi reviewed (commented):
The notebook is not reviewed.
Please review #1089

qingyun-wu commented on flaml/autogen/agent/agent.py:
```diff
@@ -30,22 +31,77 @@ def _remember(self, memory):
+                4. "name": In most cases, this field is not needed. When the role is "function", this field is needed to indicate the function name.
+            sender: sender of an Agent instance.
+        """
+        self._message_to_dict(message)
```
I think it should be `message = self._message_to_dict(message)`

kevin666aa replied:
Yes, thanks for catching it

qingyun-wu commented on flaml/autogen/agent/math_user_proxy_agent.py:
```diff
@@ -150,11 +155,20 @@ def __init__(
                 The limit only plays a role when human_input_mode is not "ALWAYS".
-            is_termination_msg (function): a function that takes a message and returns a boolean value.
-                This function is used to determine if a received message is a termination message.
+            is_termination_msg (function): a function that takes a dictionary (a message) and determine if this received message is a termination message.
```
```suggestion
            is_termination_msg (function): a function that takes a message in the form of a dictionary and returns a boolean value indicating if this received message is a termination message.
```

qingyun-wu commented on flaml/autogen/agent/user_proxy_agent.py:
```diff
@@ -33,19 +36,61 @@ def __init__(
                 The limit only plays a role when human_input_mode is not "ALWAYS".
-            is_termination_msg (function): a function that takes a message and returns a boolean value.
-                This function is used to determine if a received message is a termination message.
+            is_termination_msg (function): a function that takes a dictionary (a message) and determine if this received message is a termination message.
```
```suggestion
            is_termination_msg (function): a function that takes a message in the form of a dictionary and returns a boolean value indicating if this received message is a termination message.
```

sonichi commented on flaml/autogen/agent/user_proxy_agent.py:
```diff
@@ -90,23 +96,109 @@ def _execute_code(self, code_blocks):
+            args = json.loads(input_string)
+            return args
+        except json.JSONDecodeError:
+            return {}
```
catching this error here would prevent the feedback about the json decode error to be sent back to the assistant agent.

sonichi commented on flaml/autogen/agent/user_proxy_agent.py:
```diff
@@ -90,23 +96,109 @@ def _execute_code(self, code_blocks):
+                except Exception as e:
+                    content = f"Error: {e}"
+            else:
+                content = f"Error: Invalid arguments for function {func_name}."
```
Is it better to send back the json decode error message?

kevin666aa replied:
Yes, will revise and add tests.

sonichi commented on test/autogen/test_agent.py:
```diff
@@ -0,0 +1,32 @@
+        },
+        dummy_agent_2,
+    )  # receive a dict
+    # receive dict with wrong field, there should be no error raised
```
Why is it "wrong field"?

kevin666aa replied:
Changing to "receive dict without openai fields to be printed, such as "content", 'function_call'"

sonichi commented on flaml/autogen/agent/agent.py:
```diff
@@ -17,7 +18,7 @@ def __init__(self, name, system_message=""):
         self._memory = []
         # a dictionary of conversations, default value is list
-        self._conversations = defaultdict(list)
+        self._oai_conversations = defaultdict(list)
```
@JieyuZ2 Is this change OK?

JieyuZ2 replied:
LGTM

sonichi commented on flaml/autogen/agent/agent.py:
```diff
@@ -30,22 +31,79 @@ def _remember(self, memory):
+        message = self._message_to_dict(message)
+        # create openai message to be appended to the conversation
+        oai_message = {k: message[k] for k in ("content", "function_call", "name") if k in message}
+        assert oai_message, "Message must contain at least one of the following fields: content, function_call, name."
```
I don't think we should assert here. The message is not necessarily an oai message.
@qingyun-wu please advise.

sonichi replied:
The assert prevents the message to be non-oai message.

sonichi commented on test/autogen/test_agent.py:
```diff
@@ -0,0 +1,32 @@
+        },
+        dummy_agent_2,
+    )  # send a dict
+    try:
+        dummy_agent_1._send({"message": "hello"}, dummy_agent_2)  # send dict with wrong field
+    except AssertionError:
+        print("AssertionError caught")
+        pass
```
Problematic. See my comment in agent.py.

kevin666aa replied:
Revised.

sonichi commented on test/autogen/test_function_call.py:
```diff
@@ -60,5 +60,59 @@ def test_eval_math_responses():
     print(eval_math_responses(**arguments))
 
 
+def test_json_extraction():
+    from flaml.autogen.agent import UserProxyAgent
+
+    user = UserProxyAgent(name="test", use_docker=False)
+
+    jstr = '{\n"location": "Boston, MA"\n}'
+    assert user._format_json_str(jstr) == '{"location": "Boston, MA"}'
+    assert user._extract_args(jstr) == {"location": "Boston, MA"}
+
+    jstr = '{\n"code": "python",\n"query": "x=3\nprint(x)"}'
+    assert user._format_json_str(jstr) == '{"code": "python","query": "x=3\\nprint(x)"}'
+    assert user._extract_args(jstr) == {"code": "python", "query": "x=3\nprint(x)"}
+
+    jstr = '{"code": "a=\\"hello\\""}'
+    assert user._format_json_str(jstr) == '{"code": "a=\\"hello\\""}'
+    assert user._extract_args(jstr) == {"code": 'a="hello"'}
+
```
@JieyuZ2 could you please review this part?

sonichi commented on flaml/autogen/agent/user_proxy_agent.py:
```diff
@@ -14,6 +16,7 @@ def __init__(
         system_message="",
         work_dir=None,
         human_input_mode="ALWAYS",
+        function_map=defaultdict(callable),
```
why use `defaultdict`?

kevin666aa replied:
This is not necessary, changing to None

sonichi commented on flaml/autogen/agent/user_proxy_agent.py:
```diff
@@ -90,23 +96,109 @@ def _execute_code(self, code_blocks):
+                except Exception as e:
+                    content = f"Error: {e}"
+            else:
+                content = f"Error: Invalid arguments for function {func_name}."
```
add test

sonichi commented on flaml/autogen/agent/agent.py:
```diff
@@ -30,22 +31,79 @@ def _remember(self, memory):
+        print("\n", "-" * 80, flush=True, sep="")
+
+        # create openai message to be appended to the conversation
+        oai_message = {k: message[k] for k in ("content", "function_call", "name") if k in message}
+        oai_message["role"] = "function" if message.get("role") == "function" else "user"
+        self._oai_conversations[sender.name].append(oai_message)
```
In general, the message received is not necessarily an oai message.
@qingyun-wu

kevin666aa replied:
See comment above

kevin666aa replied:
See comment below.

kevin666aa commented:
Currently, we only extract valid oai fields and append to `self._oai_conversations`, which ensures that `self._oai_conversations` can be passed to openai directly. 

I revised this part and added a`_append_oai_message` method, that will be called in _send and _receive to construct this valid oai conversation. This ensures that only field "content", "function_call", "name" will be reserved. If none of  "content", "function_call" is seen, an empty str will be put in field "content". The role will be modified accordingly.

I remember we agreed that we can add an additional class variable to store complete messages, or to modify this  oai_conversations in the future.

sonichi commented:
> Currently, we only extract valid oai fields and append to `self._oai_conversations`, which ensures that `self._oai_conversations` can be passed to openai directly.
> 
> I revised this part and added a`_append_oai_message` method, that will be called in _send and _receive to construct this valid oai conversation. This ensures that only field "content", "function_call", "name" will be reserved. If none of "content", "function_call" is seen, an empty str will be put in field "content". The role will be modified accordingly.
> 
> I remember we agreed that we can add an additional class variable to store complete messages, or to modify this oai_conversations in the future.

My point is: don't prevent the message to be non-oai message. It's OK to maintain oai messages. But don't assume all the messages are oai messages.

kevin666aa commented:
> > Currently, we only extract valid oai fields and append to `self._oai_conversations`, which ensures that `self._oai_conversations` can be passed to openai directly.
> > I revised this part and added a`_append_oai_message` method, that will be called in _send and _receive to construct this valid oai conversation. This ensures that only field "content", "function_call", "name" will be reserved. If none of "content", "function_call" is seen, an empty str will be put in field "content". The role will be modified accordingly.
> > I remember we agreed that we can add an additional class variable to store complete messages, or to modify this oai_conversations in the future.
> 
> My point is: don't prevent the message to be non-oai message. It's OK to maintain oai messages. But don't assume all the messages are oai messages.

Got it. I removed the assertion, so that messages can still be non-openai message now. I understand that an agent might not use the oai_message at all, but communicate with other fields or other LLMs

sonichi commented on flaml/autogen/agent/agent.py:
```diff
@@ -30,22 +31,95 @@ def _remember(self, memory):
+        message = self._message_to_dict(message)
+        # create openai message to be appended to the openai conversation that can be passed to oai directly.
+        oai_message = {k: message[k] for k in ("content", "function_call", "name") if k in message}
+        if "content" not in oai_message and "function_call" not in oai_message:
```
I think when this condition is True, we shouldn't append the message to self._oai_conversations because it's not a oai message.

sonichi commented on flaml/autogen/agent/user_proxy_agent.py:
```diff
@@ -90,23 +96,99 @@ def _execute_code(self, code_blocks):
+                arguments = json.loads(input_string)
+            except json.JSONDecodeError as e:
+                arguments = None
+                content = f"Error: {e}\n You argument should follow json format."
```
What's `e` like? Is it good to wrap it with other strs?

sonichi commented on flaml/autogen/agent/user_proxy_agent.py:
```diff
@@ -90,23 +96,99 @@ def _execute_code(self, code_blocks):
+                    content = func(**arguments)
+                    is_exec_success = True
+                except Exception as e:
+                    content = f"Error: {e}"
```
Same question about the wrapping.

kevin666aa replied:
```    
# function name called is wrong or doesn't exist
    wrong_func_name = {"name": "subtract_num", "arguments": '{ "num_to_be_added": 5 }'}
    print(user._execute_function(func_call=wrong_func_name)[1]["content"])

    # arguments passed is not in correct json format
    wrong_json_format = {
        "name": "add_num",
        "arguments": '{ "num_to_be_added": 5, given_num: 10 }',
    }  # should be "given_num" with quotes
    print(user._execute_function(func_call=wrong_json_format)[1]["content"])

    # function execution error with wrong arguments passed
    wrong_args = {"name": "add_num", "arguments": '{ "num_to_be_added": 5, "given_num": 10 }'}
    print(user._execute_function(func_call=wrong_args)[1]["content"])
```
     
    Return:
```
Error: Function subtract_num not found.
Error: Expecting property name enclosed in double quotes: line 1 column 25 (char 24)
 You argument should follow json format.
Error: add_num() got an unexpected keyword argument 'given_num'
```

Looks good to me

sonichi commented on test/autogen/test_agent.py:
```diff
@@ -0,0 +1,34 @@
+
+    dummy_agent_1._send({"message": "hello"}, dummy_agent_2)  # send dict with wrong field
+
+    assert (
+        dummy_agent_1._oai_conversations["dummy_agent_2"][-1]["content"] == ""
+    ), "When the message is a dict without any openai fields, the content will be set to empty string."
```
See my comment earlier.

sonichi commented on test/autogen/test_agent.py:
```diff
@@ -0,0 +1,40 @@
+
+    assert pre_len == len(
+        dummy_agent_1._oai_conversations["dummy_agent_2"]
+    ), "When the message is not an valid openai message, it should not be appended to the oai conversation."
```
```suggestion
    ), "When the message is not a valid openai message, it should not be appended to the oai conversation."
```

sonichi reviewed (commented):
This is the final comment before approval.

kevin666aa commented on flaml/autogen/agent/math_user_proxy_agent.py:
```diff
@@ -391,7 +399,7 @@ class Config:
         extra = Extra.forbid
 
-    @root_validator()
+    @root_validator(skip_on_failure=True)
```
This is due to new version of pydantic. Changing setup.py line 87 138 to pydantic==1.10.9 would work.

sonichi replied:
Is it hard to avoid using deprecated decoration?

kevin666aa replied:
The error message says to replace with model_validator, but not sure this would work. Pydantic just updated to v2, I think we can wait.

sonichi commented:
Test failure: https://github.com/microsoft/FLAML/actions/runs/5436656931/jobs/9886665719#step:7:707

kevin666aa commented on flaml/autogen/agent/agent.py:
```diff
@@ -30,22 +31,77 @@ def _remember(self, memory):
         self._memory.append(memory)
 
-    def _send(self, message, recipient):
+    def _message_to_dict(message: Union[Dict, str]):
```
The file is outdated. Can you mark this change again?

kevin666aa commented:
> Reference in n

This error should be fixed. I accidentally hit the condition where the given language is unknown and fix that part in userproxyagent.
